Mathlib/Algebra/Category/Grp/Colimits.lean

```diff
@@ -42,220 +33,181 @@ then taking the quotient by the abelian group laws within each abelian group,
+  refine eq_of_sub_eq_zero ?_
+  erw [← (QuotientAddGroup.mk' (Relations F)).map_sub, ← AddMonoidHom.mem_ker]
+  rw [QuotientAddGroup.ker_mk']
+  simp only [DFinsupp.singleAddHom_apply]
+  exact AddSubgroup.subset_closure ⟨j, j', f, x, rfl⟩
+
+/-- (implementation detail) A morphism of commutative additive groups `Quot F →+ A`
+induces a cocone on `F` as long as the universes work out.
+-/
+@[simps]
+def toCocone [DecidableEq J] {A : Type w} [AddCommGroup A] (f : Quot F →+ A) : Cocone F where
+  pt := AddCommGrp.of A
+  ι := { app := fun j => f.comp (Quot.ι F j) }
+
+lemma Quot.desc_toCocone_desc [DecidableEq J] {A : Type w} [AddCommGroup A] (f : Quot F →+ A)
+    (hc : IsColimit c) : (hc.desc (toCocone F f)).comp (Quot.desc F c) = f := by
+  refine Quot.addMonoidHom_ext F (fun j x ↦ ?_)
+  rw [AddMonoidHom.comp_apply, ι_desc]
+  change ((c.ι.app j) ≫ hc.desc (toCocone F f)) _ = _
+  rw [hc.fac]
+  simp
+
+lemma Quot.desc_toCocone_desc_app [DecidableEq J] {A : Type w} [AddCommGroup A] (f : Quot F →+ A)
+    (hc : IsColimit c) (x : Quot F) : hc.desc (toCocone F f) (Quot.desc F c x) = f x := by
+  conv_rhs => rw [← Quot.desc_toCocone_desc F c f hc]
+  simp
 
-@[simp]
-theorem cocone_naturality {j j' : J} (f : j ⟶ j') :
-    F.map f ≫ coconeMorphism.{w} F j' = coconeMorphism F j := by
-  ext
-  apply Quot.sound
-  apply Relation.map
+/--
+If `c` is a cocone of `F` such that `Quot.desc F c` is bijective, then `c` is a colimit
+cocone of `F`.
+-/
+noncomputable def isColimit_of_bijective_desc [DecidableEq J]
+     (h : Function.Bijective (Quot.desc F c)) : IsColimit c where
+  desc s := AddCommGrp.ofHom ((Quot.desc F s).comp (AddEquiv.ofBijective
+    (Quot.desc F c) h).symm.toAddMonoidHom)
+  fac s j := by
+    ext x
+    simp only [Functor.const_obj_obj, AddEquiv.toAddMonoidHom_eq_coe, coe_comp, Function.comp_apply,
+      ofHom_apply, AddMonoidHom.coe_comp, AddMonoidHom.coe_coe]
+    conv_lhs => erw [← Quot.ι_desc F c j x]
+    rw [← AddEquiv.ofBijective_apply _ h, AddEquiv.symm_apply_apply]
+    simp only [Quot.ι_desc, Functor.const_obj_obj]
+  uniq s m hm := by
+    ext x
+    obtain ⟨x, rfl⟩ := h.2 x
+    simp only [AddEquiv.toAddMonoidHom_eq_coe, ofHom_apply, AddMonoidHom.coe_comp,
+      AddMonoidHom.coe_coe, Function.comp_apply]
+    rw [← AddEquiv.ofBijective_apply _ h, AddEquiv.symm_apply_apply]
+    suffices eq : m.comp (AddEquiv.ofBijective (Quot.desc F c) h) = Quot.desc F s by
+      rw [← eq]; rfl
+    refine Quot.addMonoidHom_ext F (fun j x ↦ ?_)
+    simp [← hm j]
+
+/-- (internal implementation) The colimit cocone of a functor `F`, implemented as a quotient of
+`DFinsupp (fun j ↦ F.obj j)`, under the assumption that said quotient is small.
+-/
+@[simps]
+noncomputable def colimitCocone [DecidableEq J] [Small.{w} (Quot.{w} F)] : Cocone F where
+  pt := AddCommGrp.of (Shrink (Quot F))
+  ι :=
+    { app := fun j =>
+        AddCommGrp.ofHom (Shrink.addEquiv.symm.toAddMonoidHom.comp (Quot.ι F j))
+      naturality := fun _ _ _ ↦ by
```

Comment:
Very recently, the following `where`-type syntax has become possible. It is advisable to use it:
```suggestion
    { app j :=
        AddCommGrp.ofHom (Shrink.addEquiv.symm.toAddMonoidHom.comp (Quot.ι F j))
      naturality _ _ _ := by
```